src/game/common/rts/academystats.cpp

```diff
@@ -13,34 +13,761 @@
+                        unsigned int power_frame = frame - m_powerFrame2;
+
+                        if (power_frame > m_powerFrame1) {
+                            m_powerFrame1 = power_frame;
+                        }
+                    } else {
+                        m_powerFrame2 = frame;
+                    }
+
+                    m_powerState = has_power;
+                }
+            }
+
+            if (Needs_Update()) {
+                Set_Needs_Update(false);
+            }
+        }
+    }
+}
+
+void AcademyStats::Record_Production(const Object *produced, const Object *producer)
+{
+    unsigned int frame = g_theGameLogic->Get_Frame();
+
+    if (produced->Is_KindOf(KINDOF_FS_SUPPLY_CENTER)) {
+        m_supplyCenterCount++;
+    }
+
+    if (produced->Is_KindOf(KINDOF_DOZER)) {
+        m_dozerCount++;
+    }
+
+    if (produced->Is_KindOf(KINDOF_INFANTRY) || produced->Is_KindOf(KINDOF_VEHICLE) || produced->Is_KindOf(KINDOF_DOZER)
+        || produced->Is_KindOf(KINDOF_HARVESTER)) {
+        unsigned int idle_frame = frame - m_idleFrame2;
+
+        if (idle_frame > m_idleFrame1) {
+            m_idleFrame1 = idle_frame;
+        }
+
+        m_idleFrame2 = frame;
+    }
+
+    if (produced->Is_KindOf(KINDOF_HARVESTER)) {
+        m_gathererCount++;
+    }
+
+    if (produced->Is_KindOf(KINDOF_HERO)) {
+        m_heroesCount++;
+    }
+
+    if (produced->Is_KindOf(KINDOF_FS_STRATEGY_CENTER)) {
+        m_builtStrategyCenter = true;
+    }
+
+    if (produced->Get_Contain() != nullptr) {
+        if (produced->Get_Contain()->Is_Tunnel_Contain()) {
+            m_usedTunnels = true;
+        }
+    }
+
+    if (produced->Is_KindOf(KINDOF_MONEY_HACKER) || produced->Is_KindOf(KINDOF_FS_BLACK_MARKET)
+        || produced->Is_KindOf(KINDOF_FS_SUPPLY_DROPZONE)) {
+        m_secondaryIncomeCount++;
+    }
+
+    if (produced->Is_KindOf(KINDOF_FS_BARRACKS) && g_theGameLogic->Get_Frame() <= 9000) {
+        m_builtBarracksOnTime = true;
+    }
+
+    if (produced->Is_KindOf(KINDOF_FS_WARFACTORY) && g_theGameLogic->Get_Frame() <= 18000) {
+        m_builtWarFactoryOnTime = true;
+    }
+
+    if (produced->Is_KindOf(KINDOF_FS_ADVANCED_TECH) && g_theGameLogic->Get_Frame() <= 27000) {
+        m_builtTechStructureOnTime = true;
+    }
+
+    if (produced->Is_KindOf(KINDOF_DISGUISER)) {
+        m_disguiserCount++;
+    }
+}
+
+void AcademyStats::Record_Upgrade(const UpgradeTemplate *upgrade, bool is_grant)
+{
+    if (upgrade->Get_Academy_Classify() == ACT_UPGRADE_RADAR) {
+        m_hasBuiltRadar = true;
+    }
+
+    if (!is_grant) {
+        m_upgradeResearchCount++;
+    }
+}
+
+void AcademyStats::Record_Special_Power_Used(const SpecialPowerTemplate *power)
+{
+    if (power->Get_Academy_Classify() == ACT_SUPERPOWER) {
+        m_supersUsedCount++;
+    }
+}
+
+void AcademyStats::Record_Income()
+{
+    unsigned int frame = g_theGameLogic->Get_Frame();
+
+    unsigned int income_frame = m_incomeFrame1 - frame;
```

Comment:
Looks different to `unsigned int power_frame = frame - m_powerFrame2;`. I wonder if there is mistake.

Reply:
This is correct.